### ISSUE TITLE
Update package to be Angular 9 compliant

### ISSUE THREAD

**Issue by vroussel35** [algolia/angular-instantsearch#713]
Hi algolia-search :)

**Describe the bug 🐛**

Installing this package with Angular v9 gives the following warn:

`Package "angular-instantsearch" has an incompatible peer dependency to "@angular/core" (requires ">=5.0.0 <8.0.0" (extended), would install "9.0.5").
Package "angular-instantsearch" has an incompatible peer dependency to "@angular/common" (requires ">=5.0.0 <8.0.0" (extended), would install "9.0.5").
Package "angular-instantsearch" has an incompatible peer dependency to "@angular/platform-browser" (requires ">=5.0.0 <8.0.0" (extended), would install "9.0.5").
Package "angular-instantsearch" has an incompatible peer dependency to "@angular/platform-browser-dynamic" (requires ">=5.0.0 <8.0.0" (extended), would install "9.0.5").`

Can you please update your peer dependencies?

Thanks

**Comment by francoischalifour** [comment]
We haven't spent time trying any InstantSearch apps with Angular 9 yet. Did everything work well so far in your experience?

**Comment by jadengis** [comment]
@francoischalifour I've been using v3.0.0-beta4 with angular 9 and so far haven't experienced any issues, but I haven't tested it with SSR.

**Comment by vswom** [comment]
Hello, I tried the SSR angular 9 version and I didn't meet any problem. However, it crashed with the 9.1 version : 'Uncaught ReferenceError: algoliasearchProxy_default is not defined'. I guess a new component is required to fit with 9.1 angular version.


tested with :   
"angular-instantsearch": "^3.0.0-beta.4",
"angular": 9.1.0

<img width="756" alt="Capture d’écran 2020-03-27 à 08 56 43" src="https://user-images.githubusercontent.com/6407566/77735439-dbedcc80-700a-11ea-8aa2-99e01fe2b528.png">

**Comment by Haroenv** [comment]
which version of `algoliasearch` did you use?

**Comment by vswom** [comment]
"algoliasearch": "3.33.0",

**Comment by Haroenv** [comment]
Could you make a reproduction on e.g. GitHub @vswom ?

**Comment by vswom** [comment]
@Haroenv,  I reproduce bug --> https://github.com/vswom/instant-search-bug

**Comment by vswom** [comment]
I added :
```
(window as any).algoliasearchProxy__default = null;
(window as any).encodeProxy__default = null;
```
in polyfills.ts .. 

it would seem to solve the problem..  Workaround.

**Comment by Haroenv** [comment]
Interesting workaround @vswom, I haven't yet found what the root cause is

**Comment by asithade** [comment]
> I added :
> 
> ```
> (window as any).algoliasearchProxy__default = null;
> (window as any).encodeProxy__default = null;
> ```
> 
> in polyfills.ts ..
> 
> it would seem to solve the problem.. Workaround.

I was just trying to solve this problem. Thank you!

**Comment by gligoran** [comment]
> Interesting workaround @vswom, I haven't yet found what the root cause is

@Haroenv I've had the same problem after updating Angular from 9.0.x to 9.1.0. I found that while `ng serve` and `ng build` for fine (unchange setup in angular.js), `ng build --prod` doesn't. This makes me think that the optimization/AoT might be treeshaking out some dependencies.

I've also made this https://github.com/gligoran/ng91-algolia2 example that shows the problem.

**Comment by MarkJPerry** [comment]
I'm also getting the deep link warnings. Add the variables to `polyfills.ts` does make the site work but the size of the bundles is now over 2MB so obviously the tree shaking is not working.

I suspect the issue is not down to the `angular-instantsearch` package but the `algoliasearch` dependency not being tree shakable.

**Comment by Haroenv** [comment]
The algoliasearch version which is used here is definitely not 2MB though, it's more like 100kb non-gzipped if somehow the non-lite version is used

**Comment by chhaymenghong** [comment]
> I added :
> 
> ```
> (window as any).algoliasearchProxy__default = null;
> (window as any).encodeProxy__default = null;
> ```
> 
> in polyfills.ts ..
> 
> it would seem to solve the problem.. Workaround.

I had to do
(window as any).process = {
  env: { DEBUG: undefined },
};
(window as any).algoliasearchProxy__default = null;
(window as any).encodeProxy__default = null;

my original issue: https://github.com/algolia/angular-instantsearch/issues/722

**Comment by chhaymenghong** [comment]
<img width="670" alt="Screen Shot 2020-04-18 at 4 20 40 AM" src="https://user-images.githubusercontent.com/6827915/79636491-dfb5d000-812c-11ea-8305-2f67cb036457.png">
<img width="934" alt="Screen Shot 2020-04-18 at 4 20 59 AM" src="https://user-images.githubusercontent.com/6827915/79636493-e0e6fd00-812c-11ea-8bb8-4ba5f94e5fe0.png">

**Comment by chhaymenghong** [comment]
as for the size, for some reason, algoliasearchh-helper is included twice? I am also not using range slider yet nouislider.js is included

**Comment by Haroenv** [comment]
make sure you only have `algoliasearch-helper@2` in your package.json, not @3. In your webpack config you should be able to alias that package with empty contents, however for more details we'll need a reproduction. How are you importing the components?

**Comment by chhaymenghong** [comment]
Hi @Haroenv. Here is a link https://github.com/chhaymenghong/Test. you can add your own appId and apiKey and run npm run build and then npm run bundle-report. Please advise on how to fix this. Thanks

**Comment by eunjae-lee** [comment]
Hi @chhaymenghong ,
Thank you for providing the repository.

`angular-instantsearch@3.0.0-beta.5` depends on `instantsearch.js@^3.5.4`.
Can you modify the instantsearch.js version like this?

```js
"instantsearch.js": "^3.5.4",
```

https://github.com/algolia/angular-instantsearch/blob/3.0.0-beta.5/package.json#L126

**Comment by chhaymenghong** [comment]
@eunjae-lee nice. Thank you. That fixed the duplicate import of algoliasearch-helper. Just one thing though. I am not using range slider, but nouislider.js is still included. Just curious whether there is a quick way to prevent that from being included in the bundle.
<img width="635" alt="Screen Shot 2020-04-21 at 3 03 19 PM" src="https://user-images.githubusercontent.com/6827915/79918459-47944100-83e1-11ea-9af3-82de800b47d1.png">

**Comment by Haroenv** [comment]
@chhaymenghong, something should be able to be done with null-loader, but I'm not familiar with how you would configure that in an angular setup https://webpack.js.org/loaders/null-loader/

**Comment by sameretto** [comment]
> > I added :
> > ```
> > (window as any).algoliasearchProxy__default = null;
> > (window as any).encodeProxy__default = null;
> > ```
> > 
> > 
> > in polyfills.ts ..
> > it would seem to solve the problem.. Workaround.
> 
> I had to do
> (window as any).process = {
> env: { DEBUG: undefined },
> };
> (window as any).algoliasearchProxy__default = null;
> (window as any).encodeProxy__default = null;
> 
> my original issue: #722

Still doesnt work with this.

**Comment by chhaymenghong** [comment]
> @chhaymenghong, something should be able to be done with null-loader, but I'm not familiar with how you would configure that in an angular setup https://webpack.js.org/loaders/null-loader/

oh ok. thank you!

**Comment by jadengis** [comment]
@chhaymenghong You should be able to use https://www.npmjs.com/package/@angular-builders/custom-webpack to add the null-loader for nouislider when you build you app.

**Comment by chhaymenghong** [comment]
@jadengis sweet work like a charm. Thank you!
In case other people want to see how it is done.

**Step 1**
install these two dev dependencies:
npm install null-loader --save-dev
npm i @angular-builders/custom-webpack

**Step 2**
Add this file at the same directory level as angular.json
extra-webpack.config.js

**Step 3**
In extra-webpack.config.js, add the following contents:

const path = require('path');
module.exports = {
  module: {
    rules: [
      {
        test: path.resolve(__dirname, 'node_modules/nouislider/distribute/nouislider.js'),
        use: 'null-loader',
      },
    ],
  },
};

**Step 4**
In angular.json,
replace the builder for "build" from  "@angular-devkit/build-angular:browser" to "@angular-builders/custom-webpack:browser"
and add this to "options"
"customWebpackConfig": {
              "path": "./extra-webpack.config.js",
              "mergeStrategies": {
                "module.rules": "prepend"
              }
            },

**Comment by sameretto** [comment]
> @Haroenv, I reproduce bug --> https://github.com/vswom/instant-search-bug
> 
> npm run build:ssr && npm run serve:ssr
> 
> <img alt="Capture d’écran 2020-03-27 à 11 33 38" width="739" src="https://user-images.githubusercontent.com/6407566/77747361-d8fcd700-701e-11ea-8f39-4dbd3ad57eb0.png">

Hi I still have an issue that is exactly this. I have tried adding what was mentioned above in the polyfills and that didnt work. any ideas?

**Comment by gligoran** [comment]
@Haroenv any update on this issue?

**Comment by Haroenv** [comment]
Sorry, we have had different priorities and haven't been able to deeply look into this issue as of now

**Comment by Keroxen** [comment]
So it doesn't work with Angular 11, right?...
![image](https://user-images.githubusercontent.com/31664852/122786266-cd996980-d2bc-11eb-8eda-896e142c1af7.png)

**Comment by Haroenv** [comment]
for the mean time you should be able to use angular-instantsearch@alpha and pass --legacy-peer-modules when installing @Keroxen. We are working on adding official support for the latest versions of angular soon

**Comment by Keroxen** [comment]
@Haroenv so I've installed these packages 
![image](https://user-images.githubusercontent.com/31664852/122792150-8d3cea00-d2c2-11eb-9b0e-8b10447d2daa.png)
but I'm getting some errors regarding instantsearch.js :(
![image](https://user-images.githubusercontent.com/31664852/122791775-33d4bb00-d2c2-11eb-88d3-65fd7f8e3a16.png)

**Comment by Haroenv** [comment]
this error seems to indicate that the InstantSearch instance is undefined, do you have a sandbox/repo to debug this on @Keroxen? It possibly could be typescript's synthetic default or esmoduleinterop options

**Comment by Keroxen** [comment]
I made my repo public, here it is https://github.com/Keroxen/licenta/tree/LIC-28-search-bar thank you for your help!

**Comment by tkrugg** [comment]
Can be closed now as v4.0 now supports 10+.
